File: unisync_android/app/src/main/java/com/anhquan/unisync/core/plugins/storage/StoragePlugin.kt
package com.anhquan.unisync.core.plugins.storage

import android.Manifest
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.content.pm.PackageManager
import android.os.Build
import android.os.Environment
import android.os.IBinder
import android.widget.Toast
import androidx.core.content.ContextCompat
import com.anhquan.unisync.core.Device
import com.anhquan.unisync.core.DeviceConnection
import com.anhquan.unisync.core.plugins.UnisyncPlugin
import com.anhquan.unisync.models.DeviceMessage
import com.anhquan.unisync.models.UnisyncFile
import com.anhquan.unisync.utils.execute
import com.anhquan.unisync.utils.fromMap
import com.anhquan.unisync.utils.getPayloadData
import com.anhquan.unisync.utils.runSingle
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Single
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream

class StoragePlugin(device: Device) : UnisyncPlugin(device, DeviceMessage.Type.STORAGE),
    ServiceConnection {
    private lateinit var serviceBinder: SftpService.SftpServiceBinder

    private object Method {
        const val START_SERVER = "start_server"
        const val STOP_SERVER = "stop_server"
        const val LIST_DIR = "list_dir"
        const val SEND_FILE = "send_file"
        const val GET_FILE = "get_file"
    }

    override fun listen(
        header: DeviceMessage.DeviceMessageHeader,
        data: Map<String, Any?>,
        payload: DeviceConnection.Payload?
    ) {
        super.listen(header, data, payload)
        when (header.method) {
            Method.START_SERVER -> {
                context.bindService(
                    Intent(context, SftpService::class.java),
                    this,
                    Context.BIND_AUTO_CREATE
                )
            }

            Method.STOP_SERVER -> {
                context.unbindService(this)
            }
        }
    }

    fun getDir(path: String): Single<List<UnisyncFile>> {
        return Single.create { emitter ->
            sendRequest(Method.LIST_DIR, mapOf("path" to path))
            events.execute {
                if (it.header.type == DeviceMessage.DeviceMessageHeader.Type.RESPONSE && it.header.method == Method.LIST_DIR) {
                    emitter.onSuccess(
                        (it.data["dir"] as List<*>).map { d ->
                            @Suppress("UNCHECKED_CAST")
                            fromMap(d as Map<String, Any?>, UnisyncFile::class.java)!!
                        }
                    )
                    return@execute true
                }
                return@execute false
            }
        }
    }

    fun getFile(file: UnisyncFile, onProgress: (Float) -> Unit) {
        if (!hasPermission) {
            Toast.makeText(
                context,
                "Please grant file access permission in home screen!",
                Toast.LENGTH_SHORT
            ).show()
            return
        }
        sendRequest(Method.GET_FILE, mapOf("path" to file.fullPath))
        events.execute {
            if (it.header.type == DeviceMessage.DeviceMessageHeader.Type.RESPONSE && it.header.method == Method.GET_FILE) {
                try {
                    val downloadsDir =
                        Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)
                    downloadsDir.mkdirs()
                    val f = File(downloadsDir, file.name)
                    val outputStream = FileOutputStream(f)
                    getPayloadData(it.payload!!, onProgress = onProgress, onDone = {
                        outputStream.close()
                        runSingle(
                            subscribeOn = AndroidSchedulers.mainThread()
                        ) {
                            Toast.makeText(
                                context,
                                "File saved to Downloads: ${file.name}",
                                Toast.LENGTH_SHORT
                            ).show()
                        }
                    }) { byteArray ->
                        outputStream.write(byteArray)
                    }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
                return@execute true
            }
            return@execute false
        }
    }

    fun sendFile(fileName: String, fileSize: Long, fileStream: InputStream, destination: String, onProgress: (Float) -> Unit) {
        sendRequest(
            Method.SEND_FILE,
            mapOf(
                "destination" to destination,
                "name" to fileName,
                "size" to fileSize,
            ),
            DeviceConnection.Payload(
                fileStream,
                fileSize
            ),
            onProgress
        )
    }

    override fun onDispose() {
        try {
            context.unbindService(this)
        } catch (_: Exception) {}
        super.onDispose()
    }

    override val requiredPermission: List<String>
        get() {
            return listOf<String>().let {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !Environment.isExternalStorageManager()) {
                    it.plus(Manifest.permission.MANAGE_EXTERNAL_STORAGE)
                } else {
                    it
                }
            }.filterNot {
                ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
            }
        }

    override fun onServiceConnected(name: ComponentName, service: IBinder) {
        serviceBinder = service as SftpService.SftpServiceBinder
        serviceBinder.informBound()
        sendResponse(
            Method.START_SERVER,
            mapOf(
                "port" to serviceBinder.port,
                "username" to serviceBinder.username,
                "password" to serviceBinder.password
            )
        )
    }

    override fun onServiceDisconnected(name: ComponentName) {
        serviceBinder.informUnbound()
    }
}

File: lib/core/plugins/storage/storage.plugin.dart
// ignore_for_file: non_constant_identifier_names

import 'package:unisync/core/device.dart';
import 'package:unisync/core/device_connection.dart';
import 'package:unisync/core/plugins/base_plugin.dart';
import 'package:unisync/core/plugins/storage/file_server.dart';
import 'package:unisync/core/plugins/storage/sftp_client.dart';
import 'package:unisync/models/device_message/device_message.model.dart';
import 'package:unisync/models/file/file.model.dart';
import 'package:unisync/utils/extensions/stream.ext.dart';
import 'package:unisync/utils/file_util.dart';
import 'package:unisync/utils/payload_handler.dart';
import 'package:unisync/utils/push_notification.dart';

class StoragePlugin extends UnisyncPlugin {
  StoragePlugin(Device device)
      : super(device, type: DeviceMessage.Type.STORAGE);

  static const _method = (
    START_SERVER: 'start_server',
    STOP_SERVER: 'stop_server',
    LIST_DIR: 'list_dir',
    SEND_FILE: 'send_file',
    GET_FILE: 'get_file',
  );

  @override
  void onReceive(
    DeviceMessageHeader header,
    Map<String, dynamic> data,
    Payload? payload,
  ) {
    super.onReceive(header, data, payload);
    if (header.method == _method.LIST_DIR) {
      _server.listDirectory(data['path']).then((value) {
        sendResponse(
          _method.LIST_DIR,
          data: {'dir': value.map((e) => e.toJson()).toList()},
        );
      });
    }
    if (header.method == _method.SEND_FILE) {
      getPayloadData(
        payload!.stream,
        size: payload.size,
        onProgress: (progress) {},
      ).then((value) {
        FileUtil.saveToDesktop(
          data['name'],
          value,
        ).whenComplete(() {
          PushNotification.showNotification(
            title: 'File saved to Desktop!',
            text: data['name'],
          );
        });
      });
    }
    if (header.method == _method.GET_FILE) {
      _server.getSize(data['path']).then((size) {
        sendResponse(
          _method.GET_FILE,
          data: {},
          payload: Payload(
            size,
            _server.read(data['path']),
          ),
        );
      });
    }
  }

  final _server = FileServer();
  UnisyncSFTPClient? _client;

  Future<void> startServer() {
    final c = completer();
    sendRequest(_method.START_SERVER);
    messages.listenCancellable((event) {
      if (event.header.type == DeviceMessageHeader.Type.RESPONSE) {
        if (event.header.method == _method.START_SERVER) {
          final port = event.data['port'] as int;
          final username = event.data['username'] as String;
          final password = event.data['password'] as String;
          _client = UnisyncSFTPClient(
            address: device.ipAddress!,
            port: port,
            username: username,
            password: password,
          );
          _client!.connect().whenComplete(() {
            complete(c, value: null);
          });
        } else if (event.header.status == DeviceMessageHeader.Status.ERROR) {
            completeError(c, error: Exception('No permission!'));
        }
        return true;
      }
      return false;
    });
    return future(c);
  }

  void stopServer() {
    sendRequest(_method.STOP_SERVER);
    _client = null;
  }

  Future<List<UnisyncFile>> list(String path) async {
    return (await _client!.list(path)).map((e) {
      return UnisyncFile(
        name: e.filename,
        type: () {
          if (e.attr.isDirectory) {
            return UnisyncFile.Type.DIRECTORY;
          }
          if (e.attr.isSymbolicLink) {
            return UnisyncFile.Type.SYMLINK;
          }
          return UnisyncFile.Type.FILE;
        }.call(),
        size: e.attr.size ?? -1,
        fullPath: '${path.replaceAll(RegExp(r' / $'), '')}/${e.filename}',
      );
    }).toList();
  }

  Future<void> put(String remoteDest, String localSource, {void Function(double)? onProgress}) =>
      _client!.put(remoteDest, localSource, onProgress: onProgress);

  Future<void> get(String remoteSource, String localDest, {void Function(double)? onProgress}) =>
      _client!.get(remoteSource, localDest, onProgress: onProgress);
}
